File: extra/healthcheck.js
/*
 * This script should be run after a period of time (180s), because the server may need some time to prepare.
 */
const { FBSD } = require("../server/util-server");

process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

let client;

const sslKey = process.env.UPTIME_KUMA_SSL_KEY || process.env.SSL_KEY || undefined;
const sslCert = process.env.UPTIME_KUMA_SSL_CERT || process.env.SSL_CERT || undefined;

if (sslKey && sslCert) {
    client = require("https");
} else {
    client = require("http");
}

// If host is omitted, the server will accept connections on the unspecified IPv6 address (::) when IPv6 is available and the unspecified IPv4 address (0.0.0.0) otherwise.
// Dual-stack support for (::)
let hostname = process.env.UPTIME_KUMA_HOST;

// Also read HOST if not *BSD, as HOST is a system environment variable in FreeBSD
if (!hostname && !FBSD) {
    hostname = process.env.HOST;
}

const port = parseInt(process.env.UPTIME_KUMA_PORT || process.env.PORT || 3001);

let options = {
    host: hostname || "127.0.0.1",
    port: port,
    timeout: 28 * 1000,
};

let request = client.request(options, (res) => {
    console.log(`Health Check OK [Res Code: ${res.statusCode}]`);
    if (res.statusCode === 302) {
        process.exit(0);
    } else {
        process.exit(1);
    }
});

request.on("error", function (err) {
    console.error("Health Check ERROR");
    process.exit(1);
});

request.end();

File: extra/healthcheck.go
/*
 * If changed, have to run `npm run build-docker-builder-go`.
 * This script should be run after a period of time (180s), because the server may need some time to prepare.
 */
package main

import (
	"crypto/tls"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"runtime"
	"strings"
	"time"
)

func main() {
	isFreeBSD := runtime.GOOS == "freebsd"

	// Is K8S + uptime-kuma as the container name
	// See #2083
	isK8s := strings.HasPrefix(os.Getenv("UPTIME_KUMA_PORT"), "tcp://")

	// process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{
		InsecureSkipVerify: true,
	}

	client := http.Client{
		Timeout: 28 * time.Second,
	}

	sslKey := os.Getenv("UPTIME_KUMA_SSL_KEY")
	if len(sslKey) == 0 {
		sslKey = os.Getenv("SSL_KEY")
	}

	sslCert := os.Getenv("UPTIME_KUMA_SSL_CERT")
	if len(sslCert) == 0 {
		sslCert = os.Getenv("SSL_CERT")
	}

	hostname := os.Getenv("UPTIME_KUMA_HOST")
	if len(hostname) == 0 && !isFreeBSD {
		hostname = os.Getenv("HOST")
	}
	if len(hostname) == 0 {
		hostname = "127.0.0.1"
	}

	port := ""
	// UPTIME_KUMA_PORT is override by K8S unexpectedly,
	if !isK8s {
		port = os.Getenv("UPTIME_KUMA_PORT")
	}
	if len(port) == 0 {
		port = os.Getenv("PORT")
	}
	if len(port) == 0 {
		port = "3001"
	}

	protocol := ""
	if len(sslKey) != 0 && len(sslCert) != 0 {
		protocol = "https"
	} else {
		protocol = "http"
	}

	url := protocol + "://" + hostname + ":" + port

	log.Println("Checking " + url)
	resp, err := client.Get(url)

	if err != nil {
		log.Fatalln(err)
	}

	defer resp.Body.Close()

	_, err = ioutil.ReadAll(resp.Body)

	if err != nil {
		log.Fatalln(err)
	}

	log.Printf("Health Check OK [Res Code: %d]\n", resp.StatusCode)

}
